Add function type information to functions in NVDAObjects/JAB/init.py

alexsimon7 opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:

Progress towards https://github.com/nvaccess/nvda/issues/13835

### Summary of the issue:

Functions in NVDAObjects/JAB/init.py lack type information.

### Description of user facing changes

N/A

### Description of development approach

Added type information to functions where type was apparent.

### Testing strategy:

N/A

### Known issues with pull request:

None known.

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced code clarity and maintainability through the addition of type annotations in key classes.
  
- **Improvements**
	- Improved type safety and readability of methods in the `JABTextInfo` and `JAB` classes, making it easier for developers to understand method parameters and return types.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alexsimon7 committed:
added type information to function header where type apparent

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -127,19 +127,19 @@ def _processHtml(text: str) -> str:
 		return offset
 
-	def _getBoundingRectFromOffset(self, offset):
+	def _getBoundingRectFromOffset(self, offset: int):
```
```suggestion
	def _getBoundingRectFromOffset(self, offset: int) -> RectLTWH:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -150,19 +150,19 @@ def _getCaretOffset(self):
 		self.obj.jabContext.setCaretPosition(offset)
 
-	def _getSelectionOffsets(self):
+	def _getSelectionOffsets(self) -> tuple:
```
```suggestion
	def _getSelectionOffsets(self) -> tuple[int, int]:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -171,15 +171,15 @@ def _getStoryLength(self):
 		return None
 
-	def _getLineOffsets(self, offset):
+	def _getLineOffsets(self, offset: int) -> tuple:
```
```suggestion
	def _getLineOffsets(self, offset: int) -> tuple[int, int]:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -193,10 +193,10 @@ def _getLineOffsets(self, offset):
 		return (start, end)
 
-	def _getParagraphOffsets(self, offset):
+	def _getParagraphOffsets(self, offset: int) -> tuple:
```
```suggestion
	def _getParagraphOffsets(self, offset: int) -> tuple[int, int]:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -216,7 +216,7 @@ def _getFormatFieldAndOffsets(self, offset, formatConfig, calculateOffsets=True)
 		return field, (offset, offset + length)
 
-	def getEmbeddedObject(self, offset=0):
+	def getEmbeddedObject(self, offset: int = 0):
```
```suggestion
	def getEmbeddedObject(self, offset: int = 0) -> JAB:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -231,7 +231,7 @@ def getEmbeddedObject(self, offset=0):
 
 class JAB(Window):
-	def findOverlayClasses(self, clsList):
+	def findOverlayClasses(self, clsList) -> None:
```
please add a type for clsList

```suggestion
	def findOverlayClasses(self, clsList: list[NVDAObject]) -> None:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -255,7 +255,7 @@ def findOverlayClasses(self, clsList):
 
 	@classmethod
-	def kwargsFromSuper(cls, kwargs, relation=None):
+	def kwargsFromSuper(cls, kwargs, relation=None) -> bool:
```
please add a type for relation

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -306,13 +306,13 @@ def _get_TextInfo(self):
 		return super(JAB, self).TextInfo
 
-	def _isEqual(self, other):
+	def _isEqual(self, other) -> bool:
```
```suggestion
	def _isEqual(self, other: Any) -> bool:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -543,10 +543,10 @@ def _get_lastChild(self):
 		return self._JABAccContextInfo.childrenCount
 
-	def _get_children(self):
+	def _get_children(self) -> list:
```
```suggestion
	def _get_children(self) -> list[JAB]:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -559,13 +559,13 @@ def _get_children(self):
 		return index
 
-	def _getJABRelationTargets(self, key):
+	def _getJABRelationTargets(self, key: str) -> list:
```
```suggestion
	def _getJABRelationTargets(self, key: str) -> list[JABHandler.JABContext]:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -714,13 +714,13 @@ def _get_table(self):
 		return self.table._jabTableInfo.jabTable.getAccessibleTableColumn(self.indexInParent) + 1
 
-	def _get_rowHeaderText(self):
+	def _get_rowHeaderText(self) -> str:
```
```suggestion
	def _get_rowHeaderText(self) -> str | None:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -742,7 +742,7 @@ def _get_rowHeaderText(self):
 			return " ".join(textList)
 
-	def _get_columnHeaderText(self):
+	def _get_columnHeaderText(self) -> str:
```
```suggestion
	def _get_columnHeaderText(self) -> str | None:
```

seanbudd reviewed (commented):
Can you please do a double-check to confirm these suggestions are right?
Please also make sure to add full type information where possible

alexsimon7 commented:
Yes. I'll do a double check on the suggestions. Can you confirm what you mean by full type information? Do you mean something like "tuple[int, int]" which you suggested? Thanks for the code review!

seanbudd commented:
yep exactly that, `tuple[int, int]`. as specific types as possible. Thanks

alexsimon7 committed:
added and double checked suggested additional type information

pre-commit-ci[bot] committed:
Pre-commit auto-fix

alexsimon7 committed:
fixed missing import statement

alexsimon7 committed:
update from master

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -127,19 +128,19 @@ def _processHtml(text: str) -> str:
 		return offset
 
-	def _getBoundingRectFromOffset(self, offset):
+	def _getBoundingRectFromOffset(self, offset: int) -> RectLTRB:
```
```suggestion
	def _getBoundingRectFromOffset(self, offset: int) -> RectLTWH:
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -7,20 +7,21 @@
-from .. import InvalidNVDAObject
 from locationHelper import RectLTWH
-
+from locationHelper import RectLTRB
```
```suggestion
```

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -559,13 +560,13 @@ def _get_children(self):
 		return index
 
-	def _getJABRelationTargets(self, key):
+	def _getJABRelationTargets(self, key: str) -> list[JABHandler.JABContext]:
```
for imports like these you may need to do code like this to avoid circular imports

https://stackoverflow.com/a/39757388/8030743

alexsimon7 committed:
changed type to RectLTRB; fixed JABHandler.JABContext type using forward reference

alexsimon7 commented:
I reverted the RectLTRB change and fixed JABHandler.JABContext using a forward reference. That fixed (locally) the failed unit tests I was experiencing before. Thanks for the help!

seanbudd commented on source/NVDAObjects/JAB/__init__.py:
```diff
@@ -255,7 +255,7 @@ def findOverlayClasses(self, clsList) -> None:
 
 	@classmethod
-	def kwargsFromSuper(cls, kwargs, relation=None) -> bool:
+	def kwargsFromSuper(cls, kwargs, relation: str = None) -> bool:
```
```suggestion
	def kwargsFromSuper(cls, kwargs, relation: str | None = None) -> bool:
```

seanbudd committed:
Update source/NVDAObjects/JAB/__init__.py

alexsimon7 commented:
@seanbudd Thanks for the help with this.
